### PR TITLE
DOCSP-38861 - OIDC

### PR DESCRIPTION
**Pull request by mongoKart**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-node/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-38861
Staging - https://preview-mongodbmongokart.gatsbyjs.io/pymongo/docsp-38861-oidc/security/authentication/

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST? (See note in PR)
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/security/authentication.txt [hunk 1]
```diff
@@ -532,10 +532,318 @@ You can set this option in two ways: by passing an argument to the
...
+
+First, use pip to install the ``azure-identity`` library, as shown in the
+following example:
+
+.. code-block:: sh
+
+   python3 -m pip install azure-identity
+
+Next, define a class that inherits from the ``OIDCCallback`` class. This class must
+implement one method, ``fetch()``, that returns the OIDC token in the form of an
```
Comment (jordan-smith721) [inline source/security/authentication.txt:738]:
```suggestion
implement a ``fetch()`` method, which returns the OIDC token in the form of an
```

File: source/security/authentication.txt [hunk 2]
```diff
@@ -532,10 +532,318 @@ You can set this option in two ways: by passing an argument to the
...
+
+GCP GKE
+~~~~~~~
+
+If your application runs on a GCP Google Kubernetes Engine (GKE) cluster with a
+`configured service account <https://cloud.google.com/kubernetes-engine/docs/how-to/service-accounts>`__,
+you can read the OIDC token from the standard service-account token-file location.
+
+First, define a class that inherits from the ``OIDCCallback`` class. This class must
+implement one method, ``fetch()``, that returns the OIDC token in the form of an
```
Comment (jordan-smith721) [inline source/security/authentication.txt:795]:
```suggestion
implement a ``fetch()`` method, which returns the OIDC token in the form of an
```

**Review by jordan-smith721 — APPROVED** [review]
Nice work on this! A few fixes but otherwise LGTM

File: source/security/authentication.txt [hunk 3]
```diff
@@ -532,10 +532,318 @@ You can set this option in two ways: by passing an argument to the
...
+
+Finally, set the following connection options:
+
+- ``authMechanism``: Set to ``"MONGODB-OIDC"``.
+- ``authMechanismProperties``: Set to the ``properties`` dictionary that you created in the
+  previous step.
+
+You can set these options in two ways: by passing arguments to the
+``MongoClient`` constructor or through parameters in your connection string.
+After you set the preceding environment variable,
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/security/authentication.txt:702]: good catch. there was so much copy/pasting for this ticket

File: source/includes/authentication/azure-imds-mongoclient.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from pymongo import MongoClient
+
+# define properties and MongoClient
+properties = {"ENVIRONMENT": "azure", "TOKEN_RESOURCE": "<audience>"}
+client = MongoClient(
+    "mongodb://<hostname>:<port>",
```
[the remark this answers is not in the document]
Comment (mongoKart) [reply source/includes/authentication/azure-imds-mongoclient.py:6]: added highlighting

File: source/includes/authentication/azure-envs-connection-string.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+client_id = "<Azure identity client ID>"
+class MyCallback(OIDCCallback):
+    def fetch(self, context: OIDCCallbackContext) -> OIDCCallbackResult:
+        credential = DefaultAzureCredential(managed_identity_client_id=client_id)
+        token = credential.get_token(f"{audience}/.default").token
+        return OIDCCallbackResult(access_token=token) 
+properties = {"OIDC_CALLBACK": MyCallback()}
+uri = ("mongodb://<hostname>:<port>/?"
+       "&authMechanism=MONGODB-OIDC"
+       "&authMechanismProperties=properties")
```
Comment (mongoKart) [inline source/includes/authentication/azure-envs-connection-string.py:16]: Can you pass properties in the connection string this way?

**Review by blink1073 — APPROVED** [review]
LGTM!

File: source/includes/authentication/azure-envs-connection-string.py (hunk 5 above)
Comment (blink1073) [reply source/includes/authentication/azure-envs-connection-string.py:16]: No, you cannot pass the callback in the connection string

**Pull request merged**